pages/security/csp.md

```diff
@@ -0,0 +1,115 @@
+
+
+#### Reporting
+CSP can also be configured to send reports to an endpoint you control when content that violates your policies is encountered.
+
+To do this, use the `report-uri` directive, passing it a fully qualified URI e.g. `https://my-public-site.gov/reports/csp`. Now, each time content from an invalid source is encountered, your browser will POST a JSON payload to the provided URI.
+
+Here is an example, from the MDN article linked in [Further Reading](#further-reader)
+
+```
+{
+  "csp-report": {
+    "document-uri": "http://example.com/signup.html",
+    "referrer": "",
+    "blocked-uri": "http://example.com/css/style.css",
+    "violated-directive": "style-src cdn.example.com",
+    "original-policy": "default-src 'none'; style-src cdn.example.com; report-uri /_/csp-reports"
+  }
+}
+```
+
+This data can of course be mixed with other request metadata, like IP address or browser User-Agent, and fed into whatever log processing system your app has in place.
+
+Reporting can only be enabled via HTTP header, not inside a `<meta>` tag!
+
+
+
+### Client-side Implementation
+To implement CSP on the client, add a `<meta>` tag to your web site's `<head>`.
+
+`<meta http-equiv="Content-Security-Policy" content="default-src 'self'; img-src https://*; child-src 'none';">`
+
+You're done!
+
+### Server-side Implementation
+Similarly, to include CSP on the server, simply return a `Content-Security-Policy: {my-policy-string}` HTTP header in your web responses.
+
+<br>
+
+Although a basic CSP policy is relatively easy to write, you might find it easier to use a server-side library to define your policies declaratively. Two of well-known and actively maintained libraries are [secureheaders](https://github.com/twitter/secureheaders), a **ruby** gem written and maintained by Twitter, and the [helmetjs](https://github.com/helmetjs/helmet) middleware for **express.js**.
+
+## Caveats
+A major caveat is that CSP, by default, prevents developers from adding inline `<style>` and `<script>` tags, and from specifying JavaScript behavior or CSS `style` attributes directly on DOM elements. This isn't a bad thing, as current best practices in front-end development discourage the use of these patterns.
+
+However, there are scenarios in which you may want to use an inline content tag, such as including a Google Analytics initialization script. For these cases, CSP allows the user to supply a `nonce` attribute to the inline script: `<script nonce=GHDGsfsd83hfdfFD3>...javascript here...</script>`. Then, in your policy, `script-src 'nonce-GHDGsfsd83hfdfFD3'`. Nonces must be regenerated on each request. Additionally, CSP 2.0 allows you to generate a SHA hash of the script content, and pass that to the `srcipt-src` directive as `'sha-{content-sha}'`. You don't need to use a `nonce` attribute if your content is hashed.
```

Comment:
It might also be worth mentioning backwards-[in]compatibility information about CORS here, e.g. what happens when a CSP 1.0-complaint browser encounters CSP 2.0 tags. My initial concern with this with CALC was that I didn't want the site to break on sites that were only 1.0 compliant, e.g. because they noticed the CSP headers and enabled CSP, but didn't understand the 2.0-style SHA hashes and ignored them, thus not allowing any scripts to load.  I found surprisingly little information on this afters searching, and ultimately found (either through reading the spec or trial and error testing, I forget which) that 1.0-only browsers essentially disable CSP as soon as they encounter something in the CSP headers that they don't understand. This means that 1.0-only browsers won't break on a CSP 2.0 site--they'll just behave the same way CSP-noncompliant browsers behave.

It might be worth mentioning this for folks who were in the same confused state as me, though I understand it's a mouthful to explain. I'm happy to submit a separate PR after this lands, if you prefer.

Reply:
I do mention in the introduction that browsers will ignore directives that aren't recognized, but I can't find references to the entire policy being ignored.

As far as I can tell, only unsupported directives are ignored, and the rest are parsed normally. CSP 2.0 is supposed to be backwards compatible with 1.0, although I would totally believe there are cases where this doesn't hold!

Reply:
Hmm, I could re-investigate if you want.  The main problem is that because simply activating CSP by default disables _everything_ and individual directives whitelist individual kinds of things, if a critical resource is made accessible only by a relatively new directive, that means that your site will break on browsers that support CSP but not that new directive.

I can try finding evidence of this being the rationale for why CSP doesn't follow the typical progressive enhancement strategy that most browser APIs follow. I really do wish that it were documented better, because it really scared me when I first started using CSP 2.

Reply:
Ah ok, I _think_ that in the case you are describing, you would want a policy that at a minimum looks like:

```
"default-src 'unsafe-inline'; script-src 'nonce-${base64}'"
```

Additionally, the user could implement the directive like this:

```
"script-src 'unsafe-inline' https: 'nonce-${base64}'"
```
In each of these cases, browsers that don't support CSP 2.0, will fall back to the `unsafe-inline` directive, allowing inline scripts to execute. That does disable a key feature of CSP, but if the browser doesn't support CSP 2.0 in the first place, inline scripts would have been able to execute anyway.

Reply:
Ohhh YES, that's it!!!! Totally. And it took me _so_ long to figure that out because it's not documented very well anywhere, and as you can see I've already forgotten what I had to do. It's also very easy to overlook, I think, since the above policy minus `unsafe-inline` would still work fine on CSP 2.0-compliant browsers and non-CSP browsers, but break utterly on CSP 1.0 browsers.

I think it would be really helpful to mention this in the guide. I'm happy to submit a PR after this lands, or if you want to, you can add it too, whatever works!

Reply:
Oh good! I was afraid I was missing some context . You're right, this behavior is definitely not intuitive. I'm not 100% certain where to include this in the guide, but I agree we should! 

If there is a good spot for it that makes sense to you (as a consumer of the guide), please do submit a PR! Otherwise, I can take a stab at figuring out where the info should live.